# This analysis takes the counts data generated from featureCounts and the metadata with the phenotype information as the input files.
# So make sure you have both files in the correct directories. In this scenario, I have copied the metadata to the main working directory (/mnt/sda1/RNA/40-815970407/Sheep)

# Load all the necessary libraries. If not installed, install using either:
install.packages("package") or
# to install biconductor packages, 
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install(version = "3.18")
BiocManager::install(c("clusterProfiler", "AnnotationHub"))

library(DESeq2)
library(dplyr)
library(tidyverse)

# set the working directory
setwd("/mnt/sda1/RNA/40-815970407/Sheep")

# create a directory to store DESeq2 outputs
#system("mkdir 6.deseq2")

# Read the raw count data generated from featureCounts
countData<-read.csv("5.featurecounts/Lambs.featurecounts.hisat2.Rmatrix",sep="\t", header=T, check.names=F)
#countData1 = countData %>% select(-contains(c("Low", "Medium")))#remove rows with Low and Medium in column headers (only for test case)
#countData = countData1
# run the below step if you want to remove any of the samples with poor mapping rates, as including this might induce noise in the deseq2 results
countData<-countData[ , !names(countData) %in% c("7085","7073")]# Remove 7085 and 7073 as they had poor mapping rates

# Remove the .bam, control, low, medium and high from the column names
colnames(countData)<-gsub("Control","",colnames(countData))
colnames(countData)<-gsub("Low","",colnames(countData))
colnames(countData)<-gsub("Medium","",colnames(countData))
colnames(countData)<-gsub("High","",colnames(countData))

orig_names <- names(countData) # keep a back-up copy of the original names
geneID <- countData$Geneid# Convert count data to a matrix of appropriate form that DEseq2 can read
countData <- as.matrix(countData[ , -1]) #removing first column geneID from the table
# make sure the rownames are gene ids and first column onwards should be samples. any other columns should be removed.otherwise deseq2 will throw error.
sampleIndex <- colnames(countData)
countData <- as.matrix(countData[,sampleIndex])
rownames(countData) <- geneID
head(countData)

# check library size (total mapped reads)
colSums(countData[,2:ncol(countData)])

# Read the metadata file
metaData <-read.csv("metadata_with_methaneinfoadded_metadata.csv",sep=",",header=T)
#metaData1 = metaData[!grepl('Low|Medium', metaData$ID),]# To remove any rows with Low or medium in the ID column. (Only for test case)
#metaData = metaData1
dim(metaData)
head(metaData)
rownames(metaData) <- metaData$ID
metaData$ID <- factor(metaData$ID)
rownames(metaData)<-gsub("[a-zA-Z ]", "", rownames(metaData))
head(metaData)

# Put the columns of the count data in the same order as rows names of the sample mapping, then make sure it worked
countData <- countData[,unique(rownames(metaData))]
all(colnames(countData) == rownames(metaData))

deseq2Data <- DESeqDataSetFromMatrix(countData=countData, colData=metaData, design= ~CH4production)

# if you have many samples, consider doing this filtering step
smallestGroupSize = 22 #total number of samples
keep <- rowSums(counts(deseq2Data) >= 10) >= smallestGroupSize
deseq2Data <- deseq2Data[keep,]

deseq2Data <- DESeq(deseq2Data)

#loop through results and extract significant DEGs for each model term
# speify the cut-offs for pval and lfc in the below variables.
# make sure to change the filenames with the cutoff values before saving the deg file (Line 89)

pval = 0.05
lfc = 0.584
results = resultsNames(deseq2Data)
upresultstable = matrix(nrow = length(results), ncol = 1, dimnames = list(results,"upDEGs"))
downresultstable = matrix(nrow = length(results), ncol = 1, dimnames = list(results,"downDEGs"))

for(i in 1:length(results)){

  res = results(deseq2Data, 
                name = results[i])
  resorder <- res[order(res$padj),]
  upDEGs = (length(na.omit(which(res$padj<pval & res$log2FoldChange > lfc))))
  downDEGs = (length(na.omit(which(res$padj<pval & res$log2FoldChange < -lfc))))
  resSig = subset(resorder, padj < pval & log2FoldChange > lfc | padj < pval & log2FoldChange < -lfc)
  write.csv(resSig , file=paste0("6.deseq2/",results[i],".0.05P.0.584LFC.updownDEGs.csv"), row.names = T)
  upresultstable[results[i],"upDEGs"] = upDEGs
  downresultstable[results[i],"downDEGs"] = downDEGs 
}

# fold change = 0.5, Log2FC= -1 (2 fold decrease)
# fc = 1.4, log2fc = 0.5 (1.5 times higher expression)
# log2fc =0 means no change



# Extract the rawcounts for these significant genes (for WGCNA)
#required_df <- countData[rownames(countData) %in% rownames(resSig1),]
#write.table(as.data.frame(required_df), '6.deseq2/Control.vs.High.sig.genes.raw.counts.csv',quote=F, row.names=TRUE)
